sympy/solvers/solveset.py

```diff
@@ -2594,18 +2676,23 @@ def linear_eq_to_matrix(equations, *symbols):
             Eq or Matrix.
             '''))
 
-    A, b = [], []
-    for i, f in enumerate(equations):
-        if isinstance(f, Equality):
-            f = f.rewrite(Add, evaluate=False)
-        coeff_list = linear_coeffs(f, *symbols)
-        b.append(-coeff_list.pop())
-        A.append(coeff_list)
-    A, b = map(Matrix, (A, b))
+    # construct the dictionaries
+    try:
+        eq, c = _linear_eq_to_dict(equations, symbols, strict=strict)
+    except PolyNonlinearError as err:
+        raise NonlinearError(str(err)) from err
+    # prepare output matrices
+    n, m = shape = len(eq), len(symbols)
+    ix = dict(zip(symbols, range(m)))
+    dat = {(row, ix[k]): d[k] for row, d in enumerate(eq) for k in d}
+    rhs = [-i for i in c]
+    del c
+    A = SparseMatrix(*shape, dat)
+    b = SparseMatrix(n, 1, rhs)
     return A, b
 
 
-def linsolve(system, *symbols):
+def linsolve(system, *symbols, strict=True):
```

Comment:
I'm still not sure how I think `linsolve` should behave.

For the main cases of interest which are things like `f(x)` and `f(x).diff(x)` I think that `linsolve` should handle those by default but I'm not sure that the flag `strict=False` really gives what we want as default behaviour for `linsolve`.

Reply:
Derivatives could be treated as independent of functions if they were masked off at the start. We could have 3 levels of strict: 
* True - no masking/error when unspecified nonlinearity is detected, 
* None - only Derivative masking, 
* False - no masking/no errors for nonlinearity in new generators